### DISCUSSION TITLE
Is it possible to use tooltips if you only have access to `HighlightedCode`

### DISCUSSION THREAD

**Discussion by C-Loftus** [code-hike/codehike#438] [Q&A] [answered]
I am using Codehike with Docusaurus. I would like to add tooltips. The example in the docs show us checking the props on the which contains a `CodeBlock` property and `tooltips`.
 
```js
async function CodeWithTooltips(props: unknown) {
  const { code, tooltips = [] } = parseProps(props, Schema)
  const highlighted = await highlight(code, theme)
```

However, in docusaurus, we seem to only have access to `HighlightedCode`.  Thus, I wasn't sure if there is any way to use tooltips in Docusaurus.

Thank you

```ts
export function MyCode({ codeblock }: { codeblock: HighlightedCode }) {
  return (
    <div>
      <Pre
        code={codeblock}
        handlers={[callout]}
        style={{ background: "#212121", color: "#fafafa" }}
      />
    </div>
  )
}
```
I referenced the docusaurus config here: 
https://github.com/code-hike/examples/blob/4f9bfc99a5d55275450c4606e9ef55b3de2afdac/with-docusaurus/src/components/code.tsx

**Comment by pomber** [comment]
Yes, you don't need to call `highlight`, the `code` is already highlighted.

Also, you need to change the `Schema` to use `HighlightedCodeBlock`

```tsx
const Schema = Block.extend({
  code: HighlightedCodeBlock,
  tooltips: z.array(Block).optional(),
})

function CodeWithTooltips(props: unknown) {
  const { code, tooltips = [] } = parseProps(props, Schema)
  const highlighted = code
  ...
```

**Reply by C-Loftus** [reply]
Ah great, I knew that with the highlight call (was just pasting for reference), but did not realize that with the schema. Thank you!

**Reply by C-Loftus** [reply]
@pomber Sorry, I am looking back at this but I am not finding this to work. It appears the `parseProps` function does not work. Am I missing something?

![image](https://github.com/user-attachments/assets/121e562c-95e7-41b9-9de6-8ed56928779f)


```ts
const Schema = Block.extend({
  code: HighlightedCodeBlock,
  tooltips: z.array(Block).optional(),
});

export function MyCode(props: unknown) {
  // Parse the incoming props using the defined schema
  const { code, tooltips = [] } = parseProps(props, Schema);
  
  code.annotations = code.annotations.map((a) => {
    const tooltip = tooltips.find((t) => t.title === a.query);
    if (!tooltip) return a;
    return {
      ...a,
      data: { ...a.data, children: tooltip.children },
    };
  });


 // ... rest is abbreviated for brevity


```
### Full component below: (still using the same docusaurus template in the codehike example)

<summary>
<details>

```ts
import { HighlightedCode, Pre } from "codehike/code";
import { callout } from "./annotations/callout";
import React from "react";
import { CopyButton } from "./copy";
import { collapse, collapseContent, collapseTrigger } from "./collapse";
import { link } from "./link";
import { Block, HighlightedCodeBlock, parseProps } from "codehike/blocks";
import { z } from "zod";

// Define the schema for the props
const Schema = Block.extend({
  code: HighlightedCodeBlock,
  tooltips: z.array(Block).optional(),
});

export function MyCode(props: unknown) {
  // Parse the incoming props using the defined schema
  const { code, tooltips = [] } = parseProps(props, Schema);
  
  code.annotations = code.annotations.map((a) => {
    const tooltip = tooltips.find((t) => t.title === a.query);
    if (!tooltip) return a;
    return {
      ...a,
      data: { ...a.data, children: tooltip.children },
    };
  });

  // Return the JSX with the code block and annotations
  return (
    <div
      style={{
        position: "relative",
        display: "inline-block",
        overflowX: "auto",
        width: "100%",
      }}
      className="px-4 bg-zinc-950 rounded"
    >
      {code.meta && (
        <div
          style={{
            backgroundColor: "#2a2d32",
            color: "#fafafa",
            paddingLeft: "1em",
            paddingTop: "0.5em",
            paddingBottom: "0.5em",
            fontStyle: "italic",
          }}
        >
          {code.meta}
        </div>
      )}

      <Pre
        code={code}
        handlers={[callout, collapse, collapseTrigger, collapseContent, link, tooltips]}
        style={{
          position: "relative",
          padding: "1em",
          background: "#2a2d32",
          color: "#fafafa",
          borderRadius: "4px",
          whiteSpace: "pre", // Prevents word wrapping
          overflowX: "auto", // Enables horizontal scrolling for the code block
          width: "100%", // Ensures the code block takes up the full width of the container
          boxSizing: "border-box", // Includes padding and border in the element's total width and height
        }}
      />
      <div
        style={{
          position: "absolute",
          top: "10px",
          right: "10px",
          zIndex: 10,
        }}
      >
        <CopyButton text={code.code} />
      </div>
    </div>
  );
}
```

</details>
</summary>

**Reply by pomber** [reply]
can you post the markdown too? or a Stackblitz reproduction would be ideal

**Reply by C-Loftus** [reply]
My markdown is similar to one of the code examples in the docs: https://codehike.org/docs/code/tooltip. It is just the following with no other content.  (escaped here to show raw md but in my code it isn't escaped)

\# Test
\```js
// !tooltip[/lorem/] description
function lorem(ipsum, dolor = 1) {
  const sit = ipsum == null ? 0 : ipsum.sit
  dolor = sit - amet(dolor)
  // !tooltip[/consectetur/] inspect
  return sit ? consectetur(ipsum) : []
}
\```

Having issues with stackblitz not replicating my node env but if it is relevant, the code can be found at https://github.com/internetofwater/docs.geoconnex.us/pull/25 where `test.md` is the markdown used, `code.tsx` is the component, and `docusaurus.config.ts` is the config where I add codehike to docusaurus.

Thank you!

**Reply by pomber** [reply]
Not sure I understand, that markdown is very different from the mdx from the [tooltip example](https://codehike.org/docs/code/tooltip), the `Schema` you are using there is for the `CodeWithTooltips` component from the example, that's a different component than `MyCode`.

**Reply by C-Loftus** [reply]
Sorry yeah I was trying to get at the fact that the highlighted code is the same. I think I am a bit confused since I thought the example in the response above was for a `HighlightedCode` component. I thought only the code was important as long as the code component is the same.

I am realizing now that I can't use shadcn due to its tailwind dependency so maybe this example isn't a good fit for me either way.

**Reply by pomber** [reply, marked as answer]
> I am realizing now that I can't use shadcn due to its tailwind dependency so maybe this example isn't a good fit for me either way.

You can use any tooltip component, or build your own, it's not tied to shadcn.

I recommend you start from the `callout` example, and convert it into a tooltip. Once you have that, try to adapt the tooltip example from the docs, which is more complex but more powerful.

**Reply by C-Loftus** [reply]
Yup that makes sense, I will take a look at that; thank you.
